src/shared/libfido2-util.c

```diff
@@ -632,22 +620,8 @@ int fido2_use_hmac_hash(
                         goto finish;
                 }
 
-                r = fido2_is_cred_in_specific_token(path, rp_id, cid, cid_size, pins, required);
-                /* We handle PIN related errors here since we have used PIN in the check.
-                 * If PIN error happens, we can avoid pin retries decreased the second time. */
-                if (IN_SET(r, -ENOANO,        /* → pin required */
-                              -ENOLCK,        /* → pin incorrect */
-                              -EOWNERDEAD)) { /* → uv blocked */
-                        /* If it's not the last device, try next one */
-                        if (i < found - 1)
-                                continue;
-
-                        /* -EOWNERDEAD is returned when uv is blocked. Map it to EAGAIN so users can reinsert
-                         * the token and try again. */
-                        if (r == -EOWNERDEAD)
-                                r = -EAGAIN;
-                        goto finish;
-                } else if (r == -ENODEV) /* not a FIDO2 device or lacking HMAC-SECRET extension */
+                r = fido2_is_cred_in_specific_token(path, rp_id, cid, cid_size, required);
+                if (r == -ENODEV) /* not a FIDO2 device or lacking HMAC-SECRET extension */
                         continue;
                 else if (r < 0)
                         log_error_errno(r, "Failed to determine whether the credential is in the token, trying anyway: %m");
```

Comment:
So I think the error handline here is a bit strange: I think the errors fido2_is_cred_in_specific_token() returns should actually all be fatal, and be propagated up. And the function should return "0" if it safely determined the token is not suitable for us, and "1" if it either is sure it is our token, or if it cannot determine it, but didn't fail.

otherwise there's gonna be a lot of "red" in the output here, because we keep logging about all kinds of errors at error level which actually we then end up handling gracefully...